runtime/include/cloe/data_broker.hpp

```diff
@@ -0,0 +1,1055 @@
+    */
+  signal_ptr_t alias(std::string_view old_name, std::string_view new_name,
+                     std::regex::flag_type f = std::regex_constants::ECMAScript) {
+    std::regex regex = std::regex(std::string(old_name), f);
+    auto it1 = signals().begin();
+    auto it2 = signals().begin();
+    auto end = signals().end();
+    const auto predicate = [&](const auto& item) -> bool {
+      std::smatch match;
+      return std::regex_match(item.first, match, regex);
+    };
+    it1 = (it1 != end) ? std::find_if(it1, end, predicate) : end;
+    it2 = (it1 != end) ? std::find_if(std::next(it1), end, predicate) : end;
+    if (it2 != end) {
+      throw std::invalid_argument(fmt::format(
+          "Regex pattern matches more than one signal. First match: '{}' Second Match '{}'",
+          it1->first, it2->first));
+    }
+    if (it1 == end) {
+      throw std::invalid_argument(
+          fmt::format("Regex pattern matches zero signals. Pattern: '{}'", old_name));
+      return nullptr;
+    }
+    signal_ptr_t result = alias(it1->second, new_name);
+    return result;
+  }
+
+  /**
+    * Declares a new signal
+    * \tparam T Type of the signal
+    * \param name Name of the signal
+    * \return Pointer to the specified signal
+    */
+  template <typename T>
+  signal_ptr_t declare(const std::string_view new_name) {
+    assert_static_type<T>();
+    using compatible_type = databroker::compatible_base_t<T>;
+
+    signal_ptr_t signal = Signal::create<compatible_type>();
+    alias(signal, new_name);
+    return signal;
+  }
+  /**
+     * Declares a new signal
+     * \tparam T Type of the signal
+     * \param name Name of the signal
+     * \return Container<T> storing the signal value
+     */
+  template <typename T>
+  [[nodiscard]] Container<databroker::compatible_base_t<T>> implement(std::string_view new_name) {
+    assert_static_type<T>();
+    using compatible_type = databroker::compatible_base_t<T>;
+
+    signal_ptr_t signal = declare<compatible_type>(new_name);
+    Container<compatible_type> container = signal->create_container<compatible_type>();
+    return container;
+  }
+  /**
+    * Returns the signal with the given name
+    * \param name Name of the signal
+    * \return Signal with the given name
+    */
+  [[nodiscard]] signal_ptr_t signal(std::string_view name) const {
+    auto iter = (*this)[name];
+    if (iter != signals_.end()) {
+      return iter->second;
+    }
+    throw std::invalid_argument(fmt::format("A signal with name '{}' was not found.", name));
+  }
+  /**
+    * Returns the signal with the given name
+    * \param name Name of the signal
+    * \return Signal with the given name
+    */
+  [[nodiscard]] signal_ptr_t signal(std::string_view name) {
+    auto iter = (*this)[name];
+    if (iter != signals_.end()) {
+      return iter->second;
+    }
+    throw std::invalid_argument(fmt::format("A signal with name '{}' was not found.", name));
```

Comment:
Isn't `out_of_range` better?

Reply:
`invalid_argument` is something that to me indicates the argument is always invalid. Here it just isn't found.

I personally think that `std::optional` is even better though. Or at least the user should have the option to not rely on exceptions for handling the case where something doesn't exist.

Reply:
Done.